waspc/packages/wasp-config/src/userApi.ts

```diff
@@ -0,0 +1,278 @@
+  fn: ExtImport
+  middlewareConfigFn?: ExtImport
+  entities: string[]
+  httpRoute: AppSpec.HttpRoute[]
+  auth?: boolean
+}
+
+export type JobConfig = {
+  executor: AppSpec.JobExecutor
+  perform: Perform
+  schedule?: ScheduleConfig
+  entities: string[]
+}
+
+export type Crud = {
+  entity: string
+  operations: CrudOperations
+}
+
+export type CrudOperations = {
+  get?: CrudOperationOptions
+  getAll?: CrudOperationOptions
+  create?: CrudOperationOptions
+  update?: CrudOperationOptions
+  delete?: CrudOperationOptions
+}
+
+export type CrudOperationOptions = {
+  isPublic?: boolean
+  overrideFn?: ExtImport
+}
+
+export type Perform = {
+  fn: ExtImport
+  executorOptions?: ExecutorOptions
+}
+
+export type ScheduleConfig = {
+  cron: string
+  args?: object
+  executorOptions?: ExecutorOptions
+}
+
+type ExecutorOptions = {
+  // TODO: Type this better and test it
+  // rewriting waspc/todoApp should make sure it works
+  pgBoss: object
+}
+
+export type QueryConfig = {
+  fn: ExtImport
+  entities: string[]
+  auth?: boolean
+}
+
+export type EmailSenderConfig = AppSpec.EmailSender
+
+export type AuthConfig = {
+  userEntity: string
+  methods: AuthMethods
+  externalAuthEntity?: string
+  onAuthFailedRedirectTo: string
+  onAuthSucceededRedirectTo?: string
+  onBeforeSignup?: ExtImport
+  onAfterSignup?: ExtImport
+  onBeforeOAuthRedirect?: ExtImport
+  onBeforeLogin?: ExtImport
+  onAfterLogin?: ExtImport
+}
+
+export type AuthMethods = {
+  usernameAndPassword?: UsernameAndPasswordConfig
+  discord?: ExternalAuthConfig
+  google?: ExternalAuthConfig
+  gitHub?: ExternalAuthConfig
+  keycloak?: ExternalAuthConfig
+  email?: EmailAuthConfig
+}
+
+export type UsernameAndPasswordConfig = {
+  userSignupFields?: ExtImport
+}
+
+export type ExternalAuthConfig = {
+  configFn?: ExtImport
+  userSignupFields?: ExtImport
+}
+
+export type EmailAuthConfig = {
+  userSignupFields?: ExtImport
+  fromField: EmailFromField
+  emailVerification: EmailVerificationConfig
+  passwordReset: PasswordResetConfig
+}
+
+export type EmailSender = {
+  provider: EmailProvider
+  defaultFrom?: EmailFromField
+}
+
+export type EmailProvider = AppSpec.EmailProvider
+
+export type EmailFromField = {
+  name?: string
+  email: string
+}
+
+export type EmailVerificationConfig = {
+  getEmailContentFn?: ExtImport
+  clientRoute: string
+}
+
+export type PasswordResetConfig = {
+  getEmailContentFn?: ExtImport
+  clientRoute: string
+}
+
+export type UserSpec = {
+  app: { name: string; config: AppConfig }
+  actions: Map<string, ActionConfig>
+  apiNamespaces: Map<string, ApiNamespaceConfig>
+  apis: Map<string, ApiConfig>
+  auth?: AuthConfig
+  client?: ClientConfig
+  cruds: Map<string, Crud>
+  db?: DbConfig
+  emailSender?: EmailSenderConfig
+  jobs: Map<string, JobConfig>
+  pages: Map<string, PageConfig>
+  queries: Map<string, QueryConfig>
+  routes: Map<string, RouteConfig>
+  server?: ServerConfig
+  websocket?: WebsocketConfig
+}
```

Comment:
So some stuff was left in app, like head, title, version, but some stuff got taken out, like auth, client, server, ... . How is that? That is in order to be able to define these separately so they can reference other stuff? We should probably figure out how to best draw the line there, so it is not just "does this reference something else or not" but instead it should also be logical from the user's side.

Reply:
> We should probably figure out how to best draw the line there, so it is not just "does this reference something else or not" **but instead it should also be logical from the user's side.**

Yeah, that's what I tried to do.

`auth` is the only part that had to be extracted due to technical limitations, but we also wanted to extract it (because it made sense for the user and the app declaration was crowded). The rest (client, server) were extracted purely because it made sense for the user.

Reply:
OK, it kind of makes sense to me, but if you had to explain it, how would you explain what goes into `app` and what is a standalone thing? How are those decisions being made? Let's say I want to add a new feature that is candidate for `app`, how will I decide?

Reply:
Let's say the rules are:
- Everything an app can have more of is its own declaration (servers, clients, databases).
- Auth has to be its own declaration because of technical limitations :new_moon_with_face: